spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java

```diff
@@ -63,4 +63,9 @@ private SparkSQLProperties() {}
   // Controls the WAP branch used for write-audit-publish workflow.
   // When set, new snapshots will be committed to this branch.
   public static final String WAP_BRANCH = "spark.wap.branch";
+
+  // Controls whether to push down Iceberg system function
+  public static final String SYSTEM_FUNC_PUSH_DOWN_ENABLED =
```

Comment:
As opposed to  `StaticInvoke`, `ApplyFunctionExpression` can not code gen. So add a parameter to control enable the converting or not.

Reply:
I don't think it is clear why this is necessary. Why is this needed?

Reply:
Ah, I see from the file below.

Looks like folding constant expressions relies on `ApplyFunctionExpression` rather than `StaticInvoke`. To get a plan with `ApplyFunctionExpression`, this inserts a wrapper. I don't think that's a good idea, no is having a property to enable this.

Why can't constant folding work with `StaticInvoke`? Why is constant folding in Iceberg rather than Spark necessary in the first place?

Reply:
The system function could be analyzed into `StaticInvoke` or `ApplyFunctionExpression`. However, it will be analyzed into `StaticInvoke` if it has a [magic method](https://github.com/apache/spark/blob/f139733b92d421233a3fd35374236ef084dfd10d/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java#L35). 

`StaticInvoke` can not be translated into a data source filter while `ApplyFunctionExpression` could. The `ApplyFunctionExpression` is for a user-defined function and will be translated into `UserDefinedScalarFunc` in filter pushing down. That's the reason I add a wrapper class to hide the magic method for the system functions. 

The config here is because `ApplyFunctionExpression` can not leverage WholeStageCodegen which could have a slight performance decrease in some cases.

The workaround solution is not the best one. The best choice is we patch spark upstream to support translating `StaticInvoke` to the data source filter or leveraging the magic method in `ApplyFunctionExpression` for some cases.

Reply:
I'm not following. Is this PR intended to fix pushdown or constant folding? It seems like it should be fixing constant folding because pushdown is working just fine with `StaticInvoke` after the last PR isn't it?

Reply:
#7886 is the first part code. It changes filters to V2 filters and adds support to convert system functions wrapped in Spark UserDefinedScalarFunc to Iceberg expressions.

Actually, I have mentioned in the description:  https://github.com/apache/iceberg/pull/7886#issue-1770844392. Maybe the description is not clearly enough at #7886.

I have tried combining them in one patch. And split into this based on the https://github.com/apache/iceberg/pull/7886#discussion_r1264197758

Reply:
Okay, I understand. Thanks for explaining. I thought this was about constant folding, but it's really two issues. `StaticInvoke` can't be converted to a v2 filter and using `ApplyFunctionExpression` instead breaks constant folding because that expression is not `foldable`.

I don't like introducing more problems, so it's unfortunate that this breaks constant folding. I also would really like to avoid adding a configuration that chooses between `StaticInvoke` and `ApplyFunctionExpression`. I think it may be cleaner to choose when to convert `StaticInvoke` rather than using a wrapper. Converting in a rule would allow us to check whether the `StaticInvoke` expression would be folded and leave it alone. We could also be more targeted in a rule and only convert `StaticInvoke` expressions that are in predicates that could be pushed down. An expression in a `Project` should be left alone.

There could be a problem with that approach, which is that Spark might apply a rule that keeps convert back to `StaticInvoke`. If needed, maybe we could inject a wrapper at that point, or create a subclass of `ApplyFunctionExpression` that doesn't match the rule.

Reply:
Thanks @rdblue for your understanding. 
>  I think it may be cleaner to choose when to convert StaticInvoke rather than using a wrapper.

This is what I did in the first commit here: https://github.com/apache/iceberg/pull/8088/commits/ba7a1440b594e67cb82ad087d7a3a5b3d5aec75a#diff-66b39085c38c19ba9906ad617b851b9dcf520901a3b56976114d9a90f0fce82e. I will change back to the implementation if it is the right direction. Some things need to be mentioned in this way:
1. We need to rebuild the system function with the `StaticInvoke` https://github.com/apache/iceberg/commit/ba7a1440b594e67cb82ad087d7a3a5b3d5aec75a#diff-ee8e519e4ad1c936c1a07aacbacf71a3975f4c352fbe7810a6a99341472a128b
2. The filter condition in a normal logical plan could be easy to handle. While there is a little complicated for row-level operation because the `Filter` is not a `child plan` in the tree. We have to align the operation in the [RowLevelCommandScanRelationPushDown](https://github.com/apache/iceberg/blob/b033de7ed0a408e542013c8c9d48ec556c8e54c6/spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/RowLevelCommandScanRelationPushDown.scala#L45) to do the conversion.

> Converting in a rule would allow us to check whether the StaticInvoke expression would be folded and leave it alone.

The constant `StaticInvoke` will be folded by the optimizer.

> There could be a problem with that approach, which is that Spark might apply a rule that keeps convert back to StaticInvoke. If needed, maybe we could inject a wrapper at that point, or create a subclass of ApplyFunctionExpression that doesn't match the rule.

As of now, I have not seen the rule for it. We could take an eye out for it.

Reply:
I went through some of those old commits over the weekend and I think that converting `StaticInvoke` when necessary is a good idea. I think we may also be able to simplify some of the rules, since the pushdown logic needs additional handling for reasons that don't necessarily limit rules for converting expressions. I'd like to see if we can simplify the rules to not have quite so many special cases.

And I also think we don't need to use the special matchers since we don't have special logic. We can probably just use the logical plans directly if need to support specific plan nodes.